Fixed #29004 -- Added inspectdb --include-views option.

bquinn opened the pull request:
https://code.djangoproject.com/ticket/29004

charettes commented:
Introspected views should be created as _unmanaged_ models, [that is `Meta.managed  = False`](https://docs.djangoproject.com/en/2.0/ref/models/options/#managed).

bquinn commented:
I completely agree, and that is indeed what happens according to my tests on sqlite3 and postgres.

But it looks like something else happens on MySQL, according to the test results above. Although oddly it works on mysql_gis.

I will investigate.

charettes commented:
> I completely agree, and that is indeed what happens according to my tests on sqlite3 and postgres.

Ahh I see. Maybe we should add a comment beside the `managed = False` when it should be kept this way because it's a view. Any thoughts on that @claudep?

> Although oddly it works on mysql_gis.

`mysql_gis` only runs the GIS related tests so it doesn't execute the code/test you've added.

bquinn commented:
I've now fixed the test so it passes on all test databases.

Re `managed = False`, that is the default for all `inspectdb`-generated models so I don't think there is a need to do anything differently for models generated from views.

ramiro commented on django/core/management/commands/inspectdb.py:
```diff
@@ -69,7 +76,21 @@ def strip_prefix(s):
-                    primary_key_column = connection.introspection.get_primary_key_column(cursor, table_name)
+                    except ValueError as e:
+                        # this might have failed because of sqlite3 handling of views
+                        table_list = connection.introspection.get_table_list(cursor)
```
You risk getting this excerpt optimized like this:
```
if any(ti.name == table_name and ti.type == 'v' for ti in connection.introspection.get_table_list(cursor)):
    constraints = {}
else:
    raise e
```
(or similar, modulo indentation) by our idiomatic code wizards :smile: . See e.g. #9493 and #9532

felixxm replied:
@ramiro .. and `else` is unnecessary :smile:

bquinn replied:
ah I knew there would be a more Pythonic way of doing that! Thanks very much. I hope I've got the indentation right.

felixxm commented on tests/inspectdb/tests.py:
```diff
@@ -267,3 +269,32 @@ def test_introspection_errors(self):
+            output = out.getvalue()
+            self.assertNotIn("class InspectdbPeopleView(models.Model):", output)
+
+            out = StringIO()
```
You don't have to initialize `out` twice. This line can be removed.

bquinn replied:
Thanks for these suggestions, it's all looking much neater now!

claudep commented on tests/inspectdb/tests.py:
```diff
@@ -4,12 +4,14 @@
 
 
-class InspectDBTestCase(TestCase):
+class InspectDBTestCase(TransactionTestCase):
```
I wouldn't transform the whole class to TransactionTestCase, but rather create a new class containing transaction-dependent tests. On some setups at least, TransactionTestCase is heavily slowering tests.

bquinn commented:
Thanks @claudep, @felixxm, @ramiro for your suggestions, I have implemented all of them.

carltongibson commented on AUTHORS:
```diff
@@ -120,6 +120,7 @@ answer newbie questions, and generally made Django that much better:
     Brandon Chinn <http://brandonchinn178.github.io>
     Brant Harris
     Brendan Hayward <brendanhayward85@gmail.com>
+    Brendan Quinn <brendan@cluefulmedia.com>
```
@bquinn You [submitted the CLA](https://www.djangoproject.com/foundation/cla/) too yes?

bquinn replied:
Yes and @freakboy3742 confirmed receipt of the signed CLA by email on 10 Jan.

carltongibson commented on docs/ref/django-admin.txt:
```diff
@@ -405,6 +405,13 @@ it because ``True`` is its default value).
+Without `--include-views`, only database tables are inspected, although views
+can be inspected individually if they are named as positional arguments to
+``inspectdb``.
+
```
Just thinking we need a 

```
.. versionadded:: 2.1
```

as [per the checklist](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#new-features)

bquinn replied:
good point! I've just checked that in. Thanks for the suggestion.

timgraham commented on django/core/management/commands/inspectdb.py:
```diff
@@ -69,7 +76,15 @@ def strip_prefix(s):
                         constraints = {}
-                    primary_key_column = connection.introspection.get_primary_key_column(cursor, table_name)
+                    except ValueError:
+                        # this might have failed because of sqlite3 handling of views
```
So this is a branch that only SQLite hits? That doesn't seem ideal -- inspectdb should be mostly database agnostic. Can that logic go in SQLite's `get_constraints()`?

bquinn replied:
I've now moved the SQLite-specific code into the SQLite backend as you suggest. It does make the main inspectdb code much neater. I've left it as a separate commit so you can check the changes, hope it helps.

timgraham commented on django/core/management/commands/inspectdb.py:
```diff
@@ -69,7 +76,15 @@ def strip_prefix(s):
+                    try:
+                        primary_key_column = connection.introspection.get_primary_key_column(cursor, table_name)
+                    except ValueError:
+                        primary_key_column = None
```
(same thing here)

bquinn commented:
I've fixed the cleanups that @felixxm suggested and have now added a comment when the model has been created from a view. Currently I just add 
`# Created from database view. Be careful when editing.`
at the top of the model. Does that cover what we need or do you have any other suggestions?

Re @charettes' original point around the `managed = False` line, that line is added for all models generated using `inspectdb` whether they are generated from tables or from views, so I'm not sure that adding the comment to that line in particular would be enough to guide developers.

Thanks for all the advice, this is a much better patch than the one I originally submitted!

charettes commented:
@bquinn my point was that Django simply can't manage views so that `managed` flag will have to remain set to `False` while other models could be changed back to `True` (or simply removed) and Django would be able to _manage_ their state changes.

bquinn commented:
Thanks @charettes, I understand. Perhaps we should add something to these lines which are added to the top of the generated models.py file?
```
#   * Remove `managed = False` lines if you wish to allow Django to create, modify, and delete the table"
```
How about if we changed that to:
````
#   * Remove `managed = False` lines if you wish to allow Django to create, modify, and delete
# the table. Views must be left as `managed = False` as Django can not manage database views.
````
Do you think that would help?

charettes commented:
@bquinn I think that would be great. Does that make sense to you @claudep?

timgraham commented:
My idea is to put `# Created from a view. Don't remove.` next to or above `managed = False`. That could be in lieu of the comment above each model.

Changing the comment at the top of the file seems okay too.

timgraham commented:
By the way, if you omit the @ mentions in the commit message, then GitHub won't spam us with notifications.

bquinn commented:
Sorry about that!

bquinn commented:
Just to be clear, in my last commit on Monday, I made all of the changes that were requested.

carltongibson commented:
Hey @bquinn. 

I think we're very close here. It's just last reviews before committing. (Measure twice, cut once and such...)

Thanks for your efforts and patience!
